chain/ethereum/src/data_source.rs

```diff
@@ -1,31 +1,38 @@
 use anyhow::{anyhow, Error};
 use anyhow::{ensure, Context};
-use ethabi::{Address, Event, Function, LogParam, ParamType, RawLog};
+use ethabi::{Address, Contract, Event, Function, LogParam, ParamType, RawLog};
 use graph::components::store::StoredDynamicDataSource;
-use graph::prelude::{Entity, SubgraphManifestValidationError};
+use graph::prelude::futures03::future::try_join;
+use graph::prelude::futures03::stream::FuturesOrdered;
+use graph::prelude::{Entity, Link, SubgraphManifestValidationError};
 use graph::slog::trace;
 use std::collections::BTreeMap;
 use std::str::FromStr;
 use std::{convert::TryFrom, sync::Arc};
-use tiny_keccak::keccak256;
-use web3::types::{Log, Transaction};
+use tiny_keccak::{keccak256, Keccak};
+use web3::types::{Log, Transaction, H256};
 
 use graph::{
     blockchain::{self, Blockchain},
     prelude::{
-        async_trait, info, serde_json, BlockNumber, CheapClone, DataSourceTemplateInfo,
-        Deserialize, EthereumCall, LightEthereumBlock, LightEthereumBlockExt, LinkResolver, Logger,
+        async_trait, info, lazy_static, serde_json, BlockNumber, CheapClone,
+        DataSourceTemplateInfo, Deserialize, EthereumCall, LightEthereumBlock,
+        LightEthereumBlockExt, LinkResolver, Logger, TryStreamExt,
     },
 };
 
-use graph::data::subgraph::{
-    BlockHandlerFilter, DataSourceContext, Mapping, MappingABI, MappingBlockHandler,
-    MappingCallHandler, MappingEventHandler, Source, TemplateSource, UnresolvedMapping,
-};
+use graph::data::subgraph::{calls_host_fn, DataSourceContext, Source};
 
 use crate::chain::Chain;
 use crate::trigger::{EthereumBlockTriggerType, EthereumTrigger, MappingTrigger};
 
+lazy_static! {
+    static ref MAX_API_VERSION: semver::Version = std::env::var("GRAPH_MAX_API_VERSION")
```

Comment:
This is here now because we may have a different set of `apiVersion`s for each chain?

Reply:
This isn't a good place to put this, I've moved it and the check to a better place in the last commit.